lib/auth/iam_join.go

```diff
@@ -0,0 +1,245 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package auth
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"crypto/rand"
+	"encoding/base64"
+	"encoding/json"
+	"io"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strings"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/utils"
+	"github.com/gravitational/teleport/lib/utils/aws"
+
+	"github.com/gravitational/trace"
+)
+
+const (
+	expectedSTSIdentityRequestBody = "Action=GetCallerIdentity&Version=2011-06-15"
+	stsHost                        = "sts.amazonaws.com"
+	challengeHeaderKey             = "X-Teleport-Challenge"
+	normalizedChallengeHeaderKey   = "x-teleport-challenge"
```

Comment:
Refer to AWS docs stating that the SigV4 headers are always lowercase?

https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-auth-using-authorization-header.html#sigv4-auth-header-overview

Reply:
done